src/stream/src/executor/source/batch_source/batch_adbc_snowflake_list.rs

```diff
@@ -0,0 +1,553 @@
+                .ok_or_else(|| {
+                    anyhow!("column_name column not found in SHOW PRIMARY KEYS result")
+                })?;
+            let key_seq_idx = schema
+                .fields()
+                .iter()
+                .position(|f| f.name().eq_ignore_ascii_case("key_sequence"))
+                .ok_or_else(|| {
+                    anyhow!("key_sequence column not found in SHOW PRIMARY KEYS result")
+                })?;
+
+            let col_name_array = batch
+                .column(col_name_idx)
+                .as_any()
+                .downcast_ref::<arrow_array_55::StringArray>()
+                .ok_or_else(|| anyhow!("column_name is not StringArray"))?;
+            let key_seq_array = batch
+                .column(key_seq_idx)
+                .as_any()
+                .downcast_ref::<arrow_array_55::Int64Array>()
+                .ok_or_else(|| anyhow!("key_sequence is not Int64Array"))?;
+
+            for i in 0..batch.num_rows() {
+                let col_name = col_name_array.value(i).to_owned();
+                let key_seq = key_seq_array.value(i);
+                pk_pairs.push((key_seq, col_name));
+            }
+        }
+
+        // Sort by key_seq to get the correct primary key column order
+        pk_pairs.sort_by_key(|(seq, _)| *seq);
+        let pk_columns: Vec<String> = pk_pairs.into_iter().map(|(_, name)| name).collect();
+
+        // Get estimated row count
+        let count_query = format!(
+            "SELECT ROW_COUNT FROM {}.INFORMATION_SCHEMA.TABLES \
+             WHERE TABLE_SCHEMA = '{}' AND TABLE_NAME = '{}'",
+            properties.database, properties.schema, properties.table
+        );
+
+        let count_batches = properties.execute_query_with_connection(connection, &count_query)?;
+
+        let mut estimated_count: i64 = 0;
+        if let Some(batch) = count_batches.first()
+            && batch.num_rows() > 0
+            && let Some(array) = batch
+                .column(0)
+                .as_any()
+                .downcast_ref::<arrow_array_55::Int64Array>()
+        {
+            estimated_count = array.value(0);
+        }
+
+        Ok((pk_columns, estimated_count))
+    }
+
+    /// Get the min and max values of the primary key column.
+    /// If time travel query fails, falls back to querying without snapshot.
+    fn get_pk_range(
+        properties: &AdbcSnowflakeProperties,
+        connection: &mut risingwave_connector::source::adbc_snowflake::Connection,
+        pk_column: &str,
+        snapshot_timestamp: Option<&str>,
+    ) -> StreamExecutorResult<(String, String)> {
+        // Try with snapshot first if available
+        if let Some(ts) = snapshot_timestamp {
+            let table_expr_with_snapshot =
+                format!("{} AT(TIMESTAMP => '{}')", properties.table_ref(), ts);
+            let range_query = format!(
+                "SELECT MIN({})::STRING, MAX({})::STRING FROM {}",
```

Comment:
This query looks like a bit heavy, do we split it in the same way in parallel cdc backfill? @zwang28